src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,503 @@
+            } else {
+                throw new NullPointerException("salt must not be null");
+            }
+            return this;
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the salt
+         * but the actual value of the salt is not yet available. An
+         * implementation should concatenate the salt into a single value once
+         * all components are available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(byte[] salt) {
+            if (salt == null) {
+                throw new NullPointerException(
+                    "salt must not be null");
+            }
+            if (salt.length != 0) {
+                return addSalt(new SecretKeySpec(salt, "Generic"));
+            } else {
+                return this;
+            }
+        }
+    }
+
+    /**
+     * Returns a builder for building {@code Extract} and
+     * {@code ExtractThenExpand} objects.
+     * <p>
+     * Note: one or more of the methods {@code addIKM} or {@code addSalt} should
+     * be called next, before calling build methods, such as
+     * {@code Builder.extractOnly()}
+     *
+     * @return a {@code Builder} to mutate
+     */
+    static Builder ofExtract() {
+        return new Builder().createBuilder();
+    }
+
+    /**
+     * Defines the input parameters of an {@code Expand} object
+     *
+     * @param prk
+     *     the pseudorandom key; must not be {@code null} in the Expand case
+     * @param info
+     *     the optional context and application specific information (may be
+     *     {@code null}); the byte[] is copied to prevent subsequent
+     *     modification
+     * @param length
+     *     the length of the output key material (must be &gt; 0 and &lt; 255 *
+     *     HMAC length)
+     *
+     * @return a new {@code Expand} object
+     *
+     * @throws NullPointerException
+     *     if {@code prk} is {@code null}
+     * @throws IllegalArgumentException
+     *     if {@code length} is not > 0
+     */
+    static Expand expandOnly(SecretKey prk, byte[] info, int length) {
+        if (prk == null) {
+            throw new NullPointerException("prk must not be null");
+        }
+        return new Expand(prk, info, length);
+    }
+
+    /**
+     * Defines the input parameters of an Extract operation as defined in <a
+     * href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.
```

Comment:
Unless there is a specific section for the case, I don't think you need to provide the link to the RFC in these 3 classes.

Reply:
Any comment?